### PR TITLE
feat(modal): allow components as content

### LINKED ISSUE #680: Modal as encapsulated component
Please remember, the issues forum is __NOT__ for support requests. It is for bugs and feature requests only.
Please read https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md and search
existing issues (both open and closed) prior to opening any new issue and ensure you follow the instructions therein.

### Bug description:
It's nice to have the Modal as Component now.

But what I'm missing or maybe don't know is how to make it as external Component.

So I got a Component with HTML and a Controller with the functions for that modal inside it.
And from another Component I then open that Modal.

What I did now was the HTML template inside the Component HTML where I wanna open the Modal.
Like the example is doing, but what I really want is an excapsulated Modal. 
What if you need to open the same Modal at 5 places?
What if you wanna lazy load the modal?

### Link to minimally-working plunker that reproduces the issue:
-

### Version of Angular, ng-bootstrap, and Bootstrap:

**Angular:**
RC5

**ng-bootstrap:**
Alpha 3

**Bootstrap:**
Alpha 3

### LINKED ISSUE #727: Modal: string argument to the open method is confusing
See http://stackoverflow.com/q/39464345/1418796

Really, the string argument is only good for unit-testing but not for any real-life applications. We should  remove it altogether or at least clarify in the docs.

On top of this we need to support components as arguments to the `open` method, see #680

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Closes #680
Closes #727

### CODE REVIEW

**Comment by colinmorelli** [comment]
How would you propose handling inputs to the components? As far as I can tell it doesn't look like this allows for them.

Another question: while I think this is a lot cleaner, is there any functional difference between passing a component to a modal, versus using the below code?

```
<template #modal>
  <my-custom-component [input]="value()"></my-custom-component>
</template>
```

I've done that for popovers, and it has worked well.

**Comment by pkozlowski-opensource** [comment]
> How would you propose handling inputs to the components? As far as I can tell it doesn't look like this allows for them.

This PR leaves out Input / Output interactions. We can add it in the (near) future if needed.

> Another question: while I think this is a lot cleaner, is there any functional difference between passing a component to a modal, versus using the below code?

The real difference is for cases where you don't have access to any template (ex. a general confirmation dialog, reusable dialogs, a sign-in service etc.). 

> I've done that for popovers, and it has worked well.

That's great, continue to use `TemplateRef`s if this works for you! We are not removing this functionality, just adding another flex point so users can use components for cases where there is no easy access to a template.

**Comment by pkozlowski-opensource** [comment]
@jnizet would be awesome if you could test-drive / review this one!

File: src/modal/modal-ref.ts [hunk 1]
```diff
@@ -3,11 +3,29 @@ import {Injectable, ComponentRef, ViewRef, ViewContainerRef} from '@angular/core
...
+   */
+  dismiss(reason?: any): void {}
+}
+
 /**
  * A reference to a newly opened modal.
  */
 @Injectable()
-export class NgbModalRef {
+export class NgbModalRef extends NgbActiveModal {
```
Comment (pkozlowski-opensource) [inline src/modal/modal-ref.ts]: Agree. I was going back and forth on this one. Somehow `NgbModalRef` and `NgbActiveModal` are very similar yet different... Will split them up.

**Comment by deepu105** [comment]
Its a good start :+1: and thank you for that. Though it addresses part of the ask, one of the very common need would be to pass inputs (more so than outputs I guess) hence would love to see support for input/output to the component. Hope it will be implemented soon

File: src/modal/modal-ref.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline src/modal/modal-ref.ts]: Done.

**Comment by pkozlowski-opensource** [comment]
> Implementing my "reusable confirm service" usecase is still not possible due to the lack of support for injected values/inputs, but we're getting much closer, now.

Yep, I will open a separate issue to track this.

**Pull request closed without merging**